File: leetcode/leetcode_cpp/insert-interval.cpp
/*
https://leetcode.com/problems/insert-interval

Solution. Add and iterate to merge
time: o(n)
space: o(n)
*/
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        if (!intervals.size()) return {newInterval};
        int n = intervals.size();
        for (int i=0; i<n; ++i) {
            auto& interval = intervals[i];
            if (interval[0] > newInterval[0]) {
                intervals.insert(intervals.begin() + i, newInterval);
                break;
            }
        }
        if (n == intervals.size())
            intervals.push_back(newInterval);

        vector<vector<int>> res;
        for (int i=0; i<intervals.size(); ++i) {
            auto& interval = intervals[i];
            if (res.empty() || res.back()[1] < interval[0]) {
                res.push_back(interval);
            }
            else {
                res.back()[1] = max(res.back()[1], interval[1]);
            }
        }
        return res;
    }
};

File: leetcode/leetcode_py/insert-interval.py
'''
https://leetcode.com/problems/insert-interval

Solution. Add and iterate to merge
time: o(n)
space: o(n)
'''
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        n = len(intervals)
        if n == 0: return [newInterval]

        for i, interval in enumerate(intervals):
            if interval[0] > newInterval[0]:
                intervals.insert(i, newInterval)
                break
        if n == len(intervals):
            intervals.append(newInterval)

        res = []
        for interval in intervals:
            if not res or res[-1][1] < interval[0]:
                res.append(interval)
            else:
                res[-1][1] = max(res[-1][1], interval[1])
        return res
            
